server/src/main/java/org/opensearch/action/search/SearchShardTask.java

```diff
@@ -46,9 +47,33 @@
  * @opensearch.internal
  */
 public class SearchShardTask extends CancellableTask {
+    // generating metadata in a lazy way since source can be quite big
+    private final Supplier<String> metadataSupplier;
 
     public SearchShardTask(long id, String type, String action, String description, TaskId parentTaskId, Map<String, String> headers) {
+        this(id, type, action, description, parentTaskId, headers, () -> "");
+    }
+
+    public SearchShardTask(
+        long id,
+        String type,
+        String action,
+        String description,
+        TaskId parentTaskId,
+        Map<String, String> headers,
+        Supplier<String> metadataSupplier
+    ) {
         super(id, type, action, description, parentTaskId, headers);
+        this.metadataSupplier = metadataSupplier;
+    }
+
+    /**
+     * Currently this method is invoked only once during unregister of task, to log the query information when the
+     * task is top resource consumer. In future if this method is going to be called more than once, the optimal
+     * approach is to memoize the result as lazily loading will result in more work if invoked more than once.
+     */
```

Comment:
Bit of a nitpick, but these types of comments almost always become stale and incorrect because it's unlikely that the comment will change whenever the external usage of `getTaskMetadata()` changes in a way that invalidates the comment.

If you're worried about this then I think you can just wrap the supplier with a [MemoizedSupplier](https://github.com/opensearch-project/OpenSearch/blob/9168f1fb43015cf7f33eab30f1300ec85f4d4305/libs/core/src/main/java/org/opensearch/common/MemoizedSupplier.java#L37) and be done with it.

Reply:
Right, I have wrapped it up with MemoizedSupplier.